packages/kit/src/runtime/app/forms.js

```diff
@@ -110,11 +113,25 @@ export function enhance(form_element, submit = () => {}) {
 	 */
 	const fallback_callback = async ({ action, result, reset }) => {
 		if (result.type === 'success') {
-			if (reset !== false) {
-				// We call reset from the prototype to avoid DOM clobbering
-				HTMLFormElement.prototype.reset.call(form_element);
-			}
-			await invalidateAll();
+			await client._invalidate_all_with_callback(() => {
+				if (reset !== false) {
+					// Emulate the behavior of form.reset() because Svelte removes the form controls' value
+					// attributes during hydration (https://github.com/sveltejs/svelte/issues/9148). If Svelte
+					// starts to keep the value attributes, we could move back to using form.reset() and simply
+					// update the form controls' value attributes after calling invalidateAll() below.
+					for (let i = 0; i < controls.length; i++) {
+						const control = controls[i];
+						if ('value' in control && control.value !== values[i]) {
```

Comment:
What happens if one of the elements is removed (or a new one added) during the invalidation process?

Reply:
Great question. The removed elements are kept around by the garbage collector until they're not referenced anymore.

But newly added elements wouldn't work with this, and I can't see a way to get the `value`s for newly added elements. I added back the original form reset call which will act as a fallback for new elements.